### ISSUE TITLE
Should new frameworks be installed without notice?

### ISSUE THREAD

**Issue by brian-kephart** [rails/rails#40639]
EDIT
tl;dr: There is no bug. There was much confusion, however, because an old app with the default configuration (`require 'rails/all'` in config/application.rb) had new frameworks added to it without any warning during the app upgrade process. Anyone reading this should probably skip to [here](https://github.com/rails/rails/issues/40639#issuecomment-729291857).
/EDIT

### Steps to reproduce
On a Rails 6.0 app that does not use Active Storage, update to 6.1.0rc1 and run `rails app:update` or `rails active_storage:update`

### Actual behavior
```
rails  active_storage:update
rails aborted!
ArgumentError: Unknown key: :source. Valid keys are: :class_name, :anonymous_class, :primary_key, :foreign_key, :dependent, :validate, :inverse_of,
:strict_loading, :autosave, :before_add, :after_add, :before_remove, :after_remove, :extend, :counter_cache, :join_table, :index_errors, :ensuring_o
wner_was
/Users/briankephart/Sites/myapp/config/environment.rb:5:in `<top (required)>'
/Users/briankephart/Sites/myapp/bin/rails:4:in `require'
/Users/briankephart/Sites/myapp/bin/rails:4:in `<top (required)>'

/Users/briankephart/Sites/myapp/bin/spring:7:in `require'
/Users/briankephart/Sites/myapp/bin/spring:7:in `block in <top (required)>'
/Users/briankephart/Sites/myapp/bin/spring:4:in `tap'
/Users/briankephart/Sites/myapp/bin/spring:4:in `<top (required)>'
/Users/briankephart/Sites/myapp/config/boot.rb:4:in `load'
/Users/briankephart/Sites/myapp/config/boot.rb:4:in `<top (required)>'
Tasks: TOP => active_storage:update => environment
(See full trace by running task with --trace)
```

### System configuration
**Rails version**: 6.0.3.4 -> 6.1.0rc1

**Ruby version**: 2.7.2

**Comment by brian-kephart** [comment]
A few more discoveries:
- This does not happen on a fresh 6.0.3.4 app made with the --skip-active-storage flag. On the fresh app, running `rails app:update` does not appear to run `rails active_storage:update`.
- It would appear that even though I've never intentionally installed Active Storage, some past upgrade on the way to 6.0.3.4 has inserted some pieces of it. I have a default storage.yml file and the migration `AddForeignKeyConstraintToActiveStorageAttachmentsForBlobId`, but no Active Storage tables in my schema.
- Deleting the config and migration files does not fix this issue.

It seems that the root of the problem is that the `rails app:update` command should not run `rails active_storage:update` at all in this case, but it is doing so anyway. My best guess is that some artifact from a previous update is causing the `app:update` task to think Active Storage is installed when it is not.

**Comment by composerinteralia** [comment]
I was not able to reproduce this on a fresh 6.0.3.4 app. Are you able to reproduce this in a fresh app? If so, could you share more detail reproduction steps?

Based on the error message, my hunch is that you may have an unnecessary `source` option being passing to a `has_many` association that is not a `through` association. This used to no-op, but after https://github.com/rails/rails/commit/b30a23f53b52e59d31358f7b80385ee5c2ba3afe it raises. Could you search your project to see if that is the case? Although I am not sure off hand why that would come up in the context of the active_storage update task...

**Comment by brian-kephart** [comment]
Thanks for looking into it @composerinteralia.

I've determined that there are multiple things happening here.

1. The app in question predates Active Storage. It had `require 'rails/all'` in config/application.rb. When Active Storage (and other things) were added to the framework, they were picked up and configured even though I never installed them. That's why I had a couple of Active Storage files, and that's why the `active_storage:update` task was being run even though I never installed Active Storage. This seems like an area where the default configuration (`require 'rails/all'`) is setting a trap for the developer. IMO, this should be addressed in the `app:update` process with a warning when new frameworks are being set up.
2. You are correct about the `source` option. In this case, the model is in a Rails engine, so I'll have to work with the engine maintainer about that. The failure had nothing to do with the `active_storage:update` task, it just happened that the class wasn't being loaded until the `active_storage:update` task ran.

I'll re-title this issue so the new frameworks issue can be discussed.

**Comment by composerinteralia** [comment]
Aha, good find! Since there is no bug in Rails, I am going to close this issue. I think the question about whether the update tasks should install new frameworks is a good one, especially given the direction we are going with options like [`--minimal`](https://github.com/rails/rails/pull/39282), but that discussion belongs on the [rails-core mailing list](https://discuss.rubyonrails.org/c/rubyonrails-core/5). Thanks!
